inference-engine/thirdparty/clDNN/src/graph_optimizer/concat_input_order.cpp

```diff
@@ -0,0 +1,223 @@
+// Copyright (c) 2020 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+
+#include "pass_manager.h"
+#include "pooling_inst.h"
+#include "convolution_inst.h"
+#include "fully_connected_inst.h"
+#include "data_inst.h"
+#include "memory_impl.h"
+#include "program_impl.h"
+
+#include <vector>
+#include <tuple>
+
+using namespace cldnn;
+
+namespace {
+
+using shuffle_range = std::pair<int32_t, int32_t>;
+
+bool can_shuffle_features(program_node& node) {
+    if (node.is_type<convolution>()) {
+        auto& conv_node = node.as<convolution>();
+        auto& wei_node = conv_node.weights();
+
+        return conv_node.get_groups() == 1 && conv_node.get_split() == 1 &&
+            conv_node.get_deformable_groups() == 1 && !conv_node.get_transposed() &&
+            !conv_node.activations_zero_points_term() &&
+            wei_node.is_type<data>() && wei_node.is_constant() && !wei_node.is_output();
+    }
+    if (node.is_type<fully_connected>()) {
+        auto& fc_node = node.as<fully_connected>();
+        auto& wei_node = fc_node.weights();
+
+        return wei_node.is_type<data>() && wei_node.is_constant() && !wei_node.is_output();
+    }
+
+    bool pass_through = false;
+    pass_through |= node.is_type<activation>();
+    pass_through |= node.is_type<pooling>();
+    // General conditions for pass-through layers
+    pass_through &= !node.is_output() && node.get_dependencies().size() == 1 && !node.has_fused_primitives();
+    if (pass_through) {
+        // Primitives that are feature order invariant, pass-through shuffled features to users
+        for (auto& user : node.get_users()) {
+            if (!can_shuffle_features(*user))
+                return false;
+        }
+        return true;
+    }
+
+    return false;
+}
+
+void shuffle_weights(data_node& node, const std::vector<shuffle_range>& ranges) {
+    // Correct for shuffled features by shuffling input feature dimension in weights.
+    // This allows to restore correct feature order on output and only changes calculation order.
+    auto wei_layout = node.get_output_layout();
+    auto& old_weights_memory = node.get_attached_memory();
+    auto new_weights_memory = old_weights_memory.get_engine()->allocate_memory(wei_layout, old_weights_memory.get_net_id(), true);
```

Comment:
Do you really need to reset this buffer? The layouts are equal, so I'd expect that all elements are initialized in the loop below

Reply:
Changed to only reset when there is padding or layout is blocked